### ISSUE TITLE
Implicits in argument position are (not) too eagerly instantiated 

### ISSUE THREAD

**Issue by nikswamy** [FStarLang/FStar#807]
```
module Test
let test = 
  let find (#a:Type) (x:nat) (l:list (int * a)) : option (int * a) = admit() in
  find 0 []
```
Fails with

```
.\test.fst(6,6-6,10) : (Error) Expected expression of type "(x:nat -> l:(list (tuple2 int a@1)) -> Tot (option (tuple2 int a@2)))"; got expression "(find )" of type "(x:nat -> l:(list (tuple2 int (?86924 find))) -> Tot (option (tuple2 int (?86924 find))))"
```

Moving `find` to the top level makes it work again.

**Comment by nikswamy** [comment]
Here's a simpler version
```
let test = 
  let find (#a:Type) (l:list a) : option a = admit() in
  find []
```

The problem really is that when type-checking the `[]`, we introduce a new unification variable "?u find", where the type of `?u` is `(#a:Type -> l:list a -> Tot (option a)) -> Tot Type`. But, when applied to `find`, we instantiate `find`'s implicit argument, giving it the type `list ?v -> Tot (option ?v)`, which is not sufficiently polymorphic.

**Comment by nikswamy** [comment]
And here's the actual problem:

```
assume val id: #a:Type0 -> a -> Tot a
assume val app_id: (#a:Type0 -> a -> Tot a) -> Tot unit
let test = app_id id
```

`.\test.fst(4,18-4,20) : (Error) Expected expression of type "(#a:Type -> uu___:a@0 -> Tot a@1)"; got expression "(id #?86699)" of type "(uu___:?86699 -> Tot ?86699)"`

**Comment by nikswamy** [comment]
This is the workaround, but it's rather annoying:

```
let test = app_id (fun (#a:Type0) -> id #a)
```

**Comment by nikswamy** [comment]
Fixed in nik_private. I use the expected type from the context, if any, to determine how many implicit arguments should be instantiated.

**Comment by mtzguido** [comment]
I think they might not be eager enough now! Consider the following artificial example:
```f#
module Bug807b

#set-options "--print_implicits"

assume val t : int -> Type
assume val i : n:int -> t n

let test () (#x:int) : t x = i x
let y : t 42 = test ()
```
```
Bug807.fst(9,15-9,22): (Error 12) Expected type "Bug807.t 42"; but "Bug807.test ()" has type "#x: Prims.int -> Prims.Tot (Bug807.t x)"
```
Instantiating by hand works perfect:
```f#
[...]
let y : t 42 = test () #_
```
```
Verified module: AA (6 milliseconds)
All verification conditions discharged successfully
```
The unit argument in `test` is needed, or this goes to `tc_value` which will work correctly. I imagine having implicits after explicits is not something we commonly do. Looking into a fix, hopefully without disturbing anything.

**Comment by mtzguido** [comment]
Managed to fix that, and *almost* everything worked in one try. There's an issue with unfolding. Basically we're computing the amount of implicits we should instantiate by looking at the **unnormalized** inferred and expected types. In this example, we're instantiating an implicit for `f 5` in the definition of `g`, since (as far as the compiler can see without normalizing) `t 5` has no implicits.

```f#
module Bug807d

let t x = #y:int -> unit -> Lemma (x + y == y + x)
assume val f : x:int -> #y:int -> unit -> Lemma (x + y == y + x)

(* Works *)
val h : unit -> #y:int -> unit -> Lemma (5 + y == y + 5)
let h () = f 5

(* Fails *)
val g : unit -> t 5
let g () = f 5
```

This came up in `CanonCommMonoid`. For now, I fixed that by the `fun #a -> ... #a` trick. Nik, maybe we can talk a bit about this next week, just to make sure I don't merge any crap. This is in `guido_tactics`, CI all green.
